Galaxus.Functional/(Option)/AsyncOptionExtensions.cs

```diff
@@ -77,4 +77,144 @@ public static async Task<T> UnwrapOrElseAsync<T>(this Task<Option<T>> self, Func
+            }
+
+            await onSome(self.Unwrap());
+        }
+        else
+        {
+            if (onNone is null)
+            {
+                throw new ArgumentNullException(nameof(onNone));
+            }
+
+            await onNone();
+        }
+    }
+
+    /// <inheritdoc cref="Option{T}.Match" />
+    public static async Task MatchAsync<T>(this Task<Option<T>> task, Func<T, Task> onSome, Func<Task> onNone)
+    {
+        var option = await task;
+        await option.MatchAsync(onSome, onNone);
+    }
+
+    /// <inheritdoc cref="Option{T}.Match" />
+    public static async Task MatchAsync<T>(this Option<T> self, Action<T> onSome, Func<Task> onNone)
+    {
+        await self.MatchAsync(async t =>
+        {
+            onSome(t);
+            await Task.CompletedTask;
+        }, onNone);
+    }
+
+    /// <inheritdoc cref="Option{T}.Match" />
+    public static async Task MatchAsync<T>(this Task<Option<T>> task, Action<T> onSome, Func<Task> onNone)
+    {
+        var option = await task;
+        await option.MatchAsync(onSome, onNone);
+    }
+
+    /// <inheritdoc cref="Option{T}.Match" />
+    public static async Task MatchAsync<T>(this Option<T> self, Func<T, Task> onSome, Action onNone)
+    {
+        await self.MatchAsync(onSome, async () =>
+        {
+            onNone();
+            await Task.CompletedTask;
+        });
+    }
+
+    /// <inheritdoc cref="Option{T}.Match" />
+    public static async Task MatchAsync<T>(this Task<Option<T>> task, Func<T, Task> onSome, Action onNone)
+    {
+        var option = await task;
+        await option.MatchAsync(onSome, onNone);
+    }
+
+    /// <inheritdoc cref="Option{T}.Match" />
+    public static async Task MatchAsync<T>(this Task<Option<T>> task, Action<T> onSome, Action onNone)
+    {
+        var option = await task;
+        option.Match(onSome, onNone);
+    }
+
+    /// <inheritdoc cref="Option{T}.Match{U}" />
+    public static async Task<U> MatchAsync<T, U>(this Option<T> self, Func<T, Task<U>> onSome, Func<Task<U>> onNone)
+    {
+        if (self.IsSome)
+        {
+            if (onSome is null)
+            {
+                throw new ArgumentNullException(nameof(onSome));
+            }
+
+            return await onSome(self.Unwrap());
+        }
+
+        if (onNone is null)
+        {
+            throw new ArgumentNullException(nameof(onNone));
+        }
+
+        return await onNone();
+    }
+
+    /// <inheritdoc cref="Option{T}.Match{U}" />
+    public static async Task<U> MatchAsync<T, U>(this Task<Option<T>> self, Func<T, Task<U>> onSome, Func<Task<U>> onNone)
+    {
+        var option = await self;
+        return await option.MatchAsync(onSome, onNone);
+    }
+
+    /// <inheritdoc cref="Option{T}.Match{U}" />
+    public static async Task<U> MatchAsync<T, U>(this Option<T> self, Func<T, U> onSome, Func<Task<U>> onNone)
+    {
+        return await self.MatchAsync(async t =>
+        {
+            await Task.CompletedTask;
```

Comment:
This one can be even simpler than the void-returning-match:

`return await self.MatchAsync(t => Task.FromResult(onSome(t)), onNone);`